src/allocator_defrag.c

```diff
@@ -0,0 +1,478 @@
+#include <stdio.h>
+#include "zmalloc.h"
+#include "serverassert.h"
+#include "allocator_defrag.h"
+
+#define UNUSED(x) (void)(x)
+
+#if defined(HAVE_DEFRAG) && defined(USE_JEMALLOC)
+
+#define STRINGIFY_(x) #x
+#define STRINGIFY(x) STRINGIFY_(x)
+
+
+#define SLAB_NFREE(out, i) out[(i) * 3]
+#define SLAB_LEN(out, i) out[(i) * 3 + 2]
+#define SLAB_NUM_REGS(out, i) out[(i) * 3 + 1]
+
+#define UTILIZATION_THRESHOLD_FACTOR_MILI (125) // 12.5% additional utilization
+
+/// @brief Helper struct to store MIB (Management Information Base) information for jemalloc bin queries.
```

Comment:
Maybe pick a name that doesn't overlap with a unit (mibibyte), even if this is well understood name in jemalloc (I didn't catch) I assumed it was the unit.

Reply:
that's how it is called in jemalloc, but I can change

Reply:
Mibibyte is MiB, not MIB. There is no collission. :)

Reply:
> Mibibyte is MiB, not MIB. There is no collission. :)

There is a case insensitive collision, which I still consider a collision.

Reply:
Fair enough. Units were never designed to be case insensitive though. m = milli, M = mega, b = bit, B = byte, etc. Being sloppy about case is a bad idea when it comes to units. :D

Reply:
will figure something... got you point

Reply:
@madolson how about "fast query key", so the `je_bin_q_helper` will be `jeBinInfoKeys` and the struct would be:
```
typedef struct jeBinInfoKeys {
    size_t curr_slabs_key[6];
    size_t curr_slabs_key_len;
    size_t nonfull_slabs_key[6];
    size_t nonfull_slabs_key_len;
    size_t curr_regs_key[6];
    size_t curr_regs_key_len;
    size_t nmalloc_key[6];
    size_t nmalloc_key_len;
    size_t ndealloc_key[6];
    size_t ndealloc_key_len;
} jeBinInfoKeys;
```